src/file-events/cpp/win_fsnotifier.cpp

```diff
@@ -330,15 +338,59 @@ void Server::runLoop(function<void(exception_ptr)> notifyStarted) {
                 break;
         }
     }
+
+    CloseHandle(threadHandle);
+}
+
+struct Command {
+    Server* server;
+    function<bool()> function;
+    condition_variable executed;
+    bool result;
+    exception_ptr failure;
+};
+
+static void CALLBACK executeOnRunLoopCallback(_In_ ULONG_PTR info) {
+    Command* command = (Command*) info;
+    try {
+        command->result = command->function();
+    } catch (const exception&) {
+        command->failure = current_exception();
+    }
+    unique_lock<mutex> lock(command->server->executionMutex);
```

Comment:
Can we only store the execution mutex in the command?

Reply:
I guess it would be simpler like that, yes. It means we need to create a new mutex per command, though that shouldn't be too frequent, and a code is somewhat nicer.

Reply:
I don't understand. My suggestion is to store a `mutex*` instead of a `Server*`.

Reply:
https://github.com/gradle/native-platform/pull/186